api/src/main/java/org/apache/iceberg/types/Types.java

```diff
@@ -412,43 +412,75 @@ public int hashCode() {
 
   public static class NestedField implements Serializable {
     public static NestedField optional(int id, String name, Type type) {
-      return new NestedField(true, id, name, type, null);
+      return new NestedField(true, id, name, type, null, null, null);
     }
 
     public static NestedField optional(int id, String name, Type type, String doc) {
-      return new NestedField(true, id, name, type, doc);
+      return new NestedField(true, id, name, type, doc, null, null);
+    }
+
+    public static NestedField optional(
```

Comment:
I thought we talked about making these package-private until the implementation was complete so that it doesn't leak into the public API before the values are read correctly?

Reply:
The reason is that I need to instantiate a `NestedField` with default value in the `TestReadDefaultValues` class, but this class lives in the iceberg-data module , which is not the same package as `types.java` in the API module.

I can annotate this method as `VisibleForTesting` and/or `Beta` , do you think that works?

Reply:
Can you move the test class? We can't merge this unless these are not visible. Annotations don't change visibility.

Reply:
It is hard to move the test class to another package since all the Avro dependencies will break in this case. I have updated the PR to make both `optional` and the `NestedField` constructor protected and inherited `NestedField` class in the test class. However, that has the risk of having to downgrade `NestedFiled` constructor to private in the future once this workaround is no longer needed. 
Another option is to keep both as private, and use reflection in the test to access them as a temporary workaround.

Reply:
kept the constructor as private and the newly introduced `NestedField` APIs as public.